rfcs/0018-transition-frontier-extensions.md

```diff
@@ -0,0 +1,101 @@
+- remove persistence diffs (and just use micro-diffs)
+- rewrite Transition Frontier to use micro-diff pattern internally for representing mutations
+- add a diff pipe for tests
+
+### Extensions Redefined
+
+A Transition Frontier Extension is an stateful, incremental view on the state of a Transiton Frontier. When a Transition Frontier is initialized, all of its extensions are also initialized using the Transition Frontier's root. Every mutation performed is represented as a list of diffs, and when the Transition Frontier updates, each Extension is notified of this list of diffs synchronously. Transition Frontier Extensions will notify the Transition Frontier if there was a update to the Extension's view when handling the diffs. If an Extension's view is updated, then a synchronous event is broadcast internally with the new view of that Extension. A Transition Frontier Extension has access to the Transition Frontier so that it can query and calculate information it requires when it handles diffs.
+
+### Extension Guidelines
+
+An extension's only input should be Transition Frontier Diffs. An extension should only be used if there is some incrementally computable view of information on top of the Transition Frontier that cannot be queried in with reasonable computational complexity on the fly. As an example, an extension which just provides the best tip is useless since the best tip can just be queried in O(1) time from the underlying Transition Frontier. An example of what makes a good extension, however, is tracking information such as the set of removed/added breadcrumbs from the best tip. With the exception of the Persistence Buffer Extension, Transiton Frontier Extensions should not resurface diffs they receives as that would be considered an abstraction leak. Diffs should (ideally) only be surfaced as part of tests. As part of this RFC, the Root\_diff extension will be removed as this extension violates this last rule, causing a leak of diff information.
+
+### New Transition Frontier Micro-diffs
+
+Transition Frontier Diffs will now be represented as smaller, composable micro-diffs rather than monolithic diffs like before. The primary advantage of this is composability, however it also helps to unify diffs with persistent diffs, allowing us to more easily implement the incremental hash computation in transition frontier and removes a layer of translation between diff formats. It also more easily allows us to defer extra computation into the Transition Frontier Extensions themselves by keeping the individual diffs light. Below is an psuedo-code implementation of the new micro-diffs.
+
+```ocaml
+type 'mutant diff =
+  | Breadcrumb_added : Breadcrumb.t -> {added: Breadcrumb.t; parent: Breadcrumb.t} diff
+  | Root_transitioned : {new: State_hash.t; garbage: Breadcrumb.t list} -> {previous: Root.t; new: Root.t; garbage: Breadcrumb.t list} diff
+  | Best_tip_changed : State_hash.t -> {previous: Breadcrumb.t; new: Breadcrumb.t} diff
```

Comment:
Should we have a `best_tip_changed` diff? I was thinking of these diffs as mutations to the underlying transition_frontier. However, computing a best_tip is actually a derived value from the transition_frontier. In my mind, it doesn't seem like this type doesn't fit in as a diff.

Reply:
The annotated ('mutant) type for the `diff` doesn't make sense to me.  I see them as some certificate representing a mutation. 

I would imagine the type of `Breadcrumb_added` and `Root_transitioned` to be the following:

```
type 'mutant diff =
  | Breadcrumb_added : Breadcrumb.t -> {added: Consensus_state.t; previous: Consensus_state.t} diff
  | Root_transitioned : {new: Root_data.t; garbage: State_hash.t list}  -> {old: Root_data.t; garbage: Consensus_state.t list } diff
```

Reply:
> Should we have a `best_tip_changed` diff? I was thinking of these diffs as mutations to the underlying transition_frontier. However, computing a best_tip is actually a derived value from the transition_frontier. In my mind, it doesn't seem like this type doesn't fit in as a diff.

Best tip is not actually a derived value with the way the transition frontier works. Yes, it can be derived, but it cannot be derived in O(1) time. However, it can be computed incrementally in O(1). Therefore, there is actually a mutation internal to the datastructure as we are changing the value of a pointer to the state hash of the best tip. Furthermore, the best tip path change and snark ref pool count extensions both rely on listening to changes to the best tip, so removing it would remove functionality we need from the extensions unless we provide more information than just the transition frontier diffs (which we are trying to avoid).

Reply:
> The annotated ('mutant) type for the `diff` doesn't make sense to me. I see them as some certificate representing a mutation.
> 
> I would imagine the type of `Breadcrumb_added` and `Root_transitioned` to be the following:
> 
> ```
> type 'mutant diff =
>   | Breadcrumb_added : Breadcrumb.t -> {added: Consensus_state.t; previous: Consensus_state.t} diff
>   | Root_transitioned : {new: Root_data.t; garbage: State_hash.t list}  -> {old: Root_data.t; garbage: Consensus_state.t list } diff
> ```

Perhaps Breadcrumb is the wrong type for representing a mutant (since the persistent transition frontier implementation does not deal with breadcrumbs), but consensus state by itself is not really the full scope of a mutation side effect either. Perhaps external transition is the right type for mutants? Since breadcrumbs are just derived from external transitions, I think that makes sense. If you just use the consensus state as the mutant, then you aren't capturing the other information belonging to the external transition such as the staged ledger diff. When folding over the mutants when computing the incremental hash, one would hash together the state hash of the external transition and a hash of the information stored in the staged ledger diff.

Reply:
> Perhaps Breadcrumb is the wrong type for representing a mutant (since the persistent transition frontier implementation does not deal with breadcrumbs), but consensus state by itself is not really the full scope of a mutation side effect either. Perhaps external transition is the right type for mutants? Since breadcrumbs are just derived from external transitions, I think that makes sense. If you just use the consensus state as the mutant, then you aren't capturing the other information belonging to the external transition such as the staged ledger diff. When folding over the mutants when computing the incremental hash, one would hash together the state hash of the external transition and a hash of the information stored in the staged ledger diff.

We can use `external_transition` as an indication of a mutation. This idea is fine with me.  I am also find having the best_tip_diff.

So in summary, I was thinking of the type to . be

```
type 'mutant diff =
  | Breadcrumb_added : Breadcrumb.t -> {added: External_transition.t; previous: External_transition.t} diff
  | Root_transitioned : {new: Root_data.t; garbage: State_hash.t list}  -> {old: Root_data.t; garbage: External_transition.t list } diff
  | Best_tip_changed : State_hash.t -> {previous: External_transition.t; new: External_transition.t} diff
```

Reply:
garbage should be a non-empty list